Evaluation Methods Documentation (#433)
Striveworks/valor: docs/metrics.md

| AP Averaged Over IOUs | The average of several AP metrics, calculated at various IOUs, grouped by class. | $$\dfrac{1}{\text{number of thresholds}} \sum\limits_{iou \in thresholds} AP_{iou}$$ |
| Mean Average Precision (mAP) 	| The mean of several AP scores, calculated over various classes. | $$\dfrac{1}{\text{number of classes}} \sum\limits_{c \in classes} AP_{c}$$ |
| mAP Averaged Over IOUs | The mean of several averaged AP scores, calculated over various classes. | $$\dfrac{1}{\text{number of thresholds}} \sum\limits_{iou \in thresholds} mAP_{iou}$$ |

## Semantic Segmentation Metrics

| Name | Description | Equation |
| :- | :- | :- |
| Intersection Over Union (IOU) | A ratio between the groundtruth and predicted regions of an image, measured as a percentage, grouped by class. |$$\dfrac{area( prediction \cap groundtruth )}{area( prediction \cup groundtruth )}$$ |
| Mean IOU 	| The average of IOUs, calculated over several different classes. | $$\dfrac{1}{\text{number of classes}} \sum\limits_{c \in classes} IOU_{c}$$ |

> # Notes

Author: Trying out github's markdown footnote feature

1. When calculating IOUs for object detection metrics, Valor handles the necessary conversion between different types of geometric annotations. For example, if your model prediction is a polygon and your groundtruth is a raster, then the raster will be converted to a polygon prior to calculating the IOU.

> # Appendix: Metric Calculations

Reviewer: overall I really like this new flow, appreciate you making these changes!

## Binary ROC AUC

### Receiver Operating Characteristic (ROC)

An ROC curve plots TPR vs. FPR at different confidence thresholds.

See [Classification: ROC Curve and AUC](https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc) for more information.

#### Determining the rate of correct predictions.

| Element | Description |
| ------- | ------------ |
| True Positive (TP) | Prediction confidence score >= threshold and is correct. |
| False Positive (FP) | Prediction confidence score >= threshold and is incorrect. |
| True Negative (TN) | Prediction confidence score < threshold and is correct. |
| False Negative (FN) | Prediction confidence score < threshold and is incorrect. |

- $\text{True Positive Rate (TPR)} = \dfrac{|TP|}{|TP| + |FN|}$

- $\text{False Positive Rate (FPR)} = \dfrac{|FP|}{|FP| + |TN|}$

These rates vary depending on a threshold so we can add a threshold parameter to be more explicit.

$$
TPR(threshold) = \dfrac{|TP(threshold)|}{|TP(threshold)| + |FN(threshold)|}
$$

$$
FPR(threshold) = \dfrac{|FP(threshold)|}{|FP(threshold)| + |TN(threshold)|}
$$

> In Valor, we use the confidence scores sorted in decreasing order as our thresholds.

Reviewer: I was confused by what "threshold" meant in the above calculations until I read this sentence. maybe move this sentence up under the header, and change it to: 

"In Valor, we use the confidence scores sorted in decreasing order as our thresholds. Using these thresholds, we can calculate our TPR and FPR as follows:"

$$
Point(score) = (FPR(score), \ TPR(score))
$$

> ### Area under the ROC curve (ROC AUC)

Reviewer: capitalize the first letters in this header to be consistent with other headers?

From the ROC curve we calculate the ROC AUC metric by approximating the integral using the trapezoidal rule formula.

$$
ROC AUC =  \sum_{i=1}^{|scores|} \frac{  \lVert Point(score_{i-1}) - Point(score_i) \rVert }{2}
$$

## Average Precision (AP)

For object-detection and instance segmentation tasks, average-precision is calculated from the intersection-over-union (IOU) of geometric predictions and ground truths.

### Multiclass Precision and Recall

Tasks that predict geometries (such as object detection or instance segmentation) use the ratio intersection-over-union (IOU) to calculate precision and recall. IOU is the ratio of intersecting area over the joint area spanned by the two geometries and is defined in the following equation.

$$Intersection \ over \ Union \ (IOU) = \dfrac{Area( prediction \cap groundtruth )}{Area( prediction \cup groundtruth )}$$

> By setting a threshold we can use IOU to determine whether we count a prediction-groundtruth pairing based on their overlap.

Reviewer: using "GroundTruth" to reference the class is fine

| Case | Description |
| :- | :- |
| True Positive (TP) | Prediction-GroundTruth pair exists with IOU >= threshold. |
| False Positive (FP) | Prediction-GroundTruth pair exists with IOU < threshold. |
| True Negative (TN) | Unused in multi-class evaluation.

> | False Negative (FN) | No Prediction exists for the GroundTruth. |

Reviewer: ```suggestion
| False Negative (FN) | No Prediction with a particular label exists for the GroundTruth. |
```

    predictions: list,
) -> list[float]:
    """Ranks ious by unique pairings."""

    retval = []
    groundtruths = set(groundtruths)
    for prediction in sorted(predictions, key=lambda x : -x.score):
        groundtruth = max(groundtruths, key=lambda x : calculate_iou(groundtruth, prediction))
        groundtruths.remove(groundtruth)
        retval.append(calculate_iou(groundtruth, prediction))
```

### Precision-Recall Curve

We can now compute the precision-recall curve using our previously ranked IOU's. We do this by iterating through the ranked IOU's and creating points cumulatively using recall and precision.

```python
def create_precision_recall_curve(
    number_of_groundtruths: int,
    ranked_ious: list[float],
    threshold: float
) -> list[tuple[float, float]]:
    """Creates the precision-recall curve from a list of IOU's and a threshold."""

    retval = []
    count_tp = 0
    for i in range(ranked_ious):
        if ranked_ious[i] >= threshold:
            count_tp += 1
        precision = count_tp / (i + 1)
        recall = count_tp / number_of_groundtruths
        retval.append((recall, precision))
```

### Calculating Average Precision

> Average precision is defined as the integration of the precision-recall curve. However, due to the varying nature of datasets it has been shown that interpolating this curve with a fixed set of points helps to reduce inconsistencies between dataset splits. The defacto standard has been to use a 101-point interpolation of the precision-recall curve to compute this integral.

Reviewer: "However, due to the varying nature of datasets " can you clarify what this means?

> Average precision is defined as the integration of the precision-recall curve. However, due to the varying nature of datasets it has been shown that interpolating this curve with a fixed set of points helps to reduce inconsistencies between dataset splits. The defacto standard has been to use a 101-point interpolation of the precision-recall curve to compute this integral.

Author: Rewrote this section

## Binary ROC AUC

### Receiver Operating Characteristic (ROC)

An ROC curve plots the True Positive Rate (TPR) vs. the False Positive Rate (FPR) at different confidence thresholds.

In Valor, we use the confidence scores sorted in decreasing order as our thresholds. Using these thresholds, we can calculate our TPR and FPR as follows:

#### Determining the Rate of Correct Predictions.

| Element | Description |
| ------- | ------------ |
| True Positive (TP) | Prediction confidence score >= threshold and is correct. |
| False Positive (FP) | Prediction confidence score >= threshold and is incorrect. |
| True Negative (TN) | Prediction confidence score < threshold and is correct. |
| False Negative (FN) | Prediction confidence score < threshold and is incorrect. |

- $\text{True Positive Rate (TPR)} = \dfrac{|TP|}{|TP| + |FN|}$

- $\text{False Positive Rate (FPR)} = \dfrac{|FP|}{|FP| + |TN|}$

As the confidence threshold varies we need to rewrite the TPR and FPR definitions to be more explicit.

$$
TPR(threshold) = \dfrac{|TP(threshold)|}{|TP(threshold)| + |FN(threshold)|}
$$

$$
FPR(threshold) = \dfrac{|FP(threshold)|}{|FP(threshold)| + |TN(threshold)|}
$$

> We now use the confidence scores sorted in decreasing order as our thresholds to generate points on a curve.

Reviewer: ```suggestion
We now use the confidence scores, sorted in decreasing order, as our thresholds in order to generate points on a curve.
```

| Precision | The number of true positives divided by the total number of positive predictions (i.e., the number of true positives plus the number of false positives). | $$\dfrac{\|TP\|}{\|TP\|+\|FP\|}$$ |
| Recall | The number of true positives divided by the total count of the class of interest (i.e., the number of true positives plus the number of true negatives). | $$\dfrac{\|TP\|}{\|TP\|+\|FN\|}$$ |
| F1 | A weighted average of precision and recall. | $$\frac{2 * Precision * Recall}{Precision + Recall}$$ |
| Accuracy | The number of true predictions divided by the total number of predictions. | $$\dfrac{\|TP\|+\|TN\|}{\|TP\|+\|TN\|+\|FP\|+\|FN\|}$$ |
| ROC AUC | The area under the Receiver Operating Characteristic (ROC) curve for the predictions generated by a given model. | See [ROCAUC methods](#roc-auc-for-classification). |

## Object Detection and Instance Segmentation Metrics

| Name | Description | Equation |
| :- | :- | :- |
| Average Precision (AP) | The weighted mean of precisions achieved at several different recall thresholds for a single Intersection over Union (IoU), grouped by class. | See [AP methods](#average-precision-ap-for-object-detection). |

> | AP Averaged Over IoUs | The average of several AP metrics, calculated at various IoUs, grouped by class. | $$\dfrac{1}{\|thresholds\|} \sum\limits_{iou \in thresholds} AP_{iou}$$ |

Reviewer: can we change "thresholds" to "number of IOU thresholds"?

can we make sure IOU is capitalized everywhere?

Author: IoU is the correct format i believe

Author: Fixed!

> | Mean Average Precision (mAP) 	| The mean of several AP scores, calculated over various classes. | $$\dfrac{1}{\|classes\|} \sum\limits_{c \in classes} AP_{c}$$ |

Reviewer: classes -> number of classes?

Author: the vertical bars signify a quantity

Reviewer 2: i'm familiar with vertical bars as cardinality but can't hurt to define it once before using in this way

> | mAP Averaged Over IoUs | The mean of several averaged AP scores, calculated over various classes. | $$\dfrac{1}{\|thresholds\|} \sum\limits_{iou \in thresholds} mAP_{iou}$$ |

Reviewer: thresholds -> number of thresholds

Author: the vertical bars signify a quantity

## Semantic Segmentation Metrics

| Name | Description | Equation |
| :- | :- | :- |
| Intersection Over Union (IoU) | A ratio between the groundtruth and predicted regions of an image, measured as a percentage, grouped by class. |$$\dfrac{area( prediction \cap groundtruth )}{area( prediction \cup groundtruth )}$$ |

> | Mean IoU 	| The average of IoUs, calculated over several different classes. | $$\dfrac{1}{\|classes\|} \sum\limits_{c \in classes} IoU_{c}$$ |

Reviewer: "classes" in divisor -> "number/count of classes"?

Author: the vertical bars signify a quantity

Reviewer: see above

> # Methods

Reviewer: you're grouping these "methods" by metrics, so this naming seems odd. what about calling this "Detailed Metric Calculations" or "Appendix: Detailed Metric Calculations"?

## ROC AUC for Classification

> ### Determining Binary Truth

Reviewer: nit: is "binary truth" the correct phrasing here? it reads awkwardly, but maybe it's a technical term I'm not aware of

| Element | Description |
| ------- | ------------ |

> | True Positive (TP) | Prediction confidence is above a threshold and is correct. |

Author: Classification doesnt use Iou

> | True Positive (TP) | Prediction confidence is above a threshold and is correct. |

Reviewer: you're right. clarify what you mean by "threshold" here

| False Positive (FP) | Prediction confidence is above a threshold and is incorrect. |
| True Negative (TN) | Prediction confidence is below a threshold and is correct. |
| False Negative (FN) | Prediction confidence is below a threshold and is incorrect. |

- $\text{True Positive Rate (TPR)} = \dfrac{|TP|}{|TP| + |FN|}$

- $\text{False Positive Rate (FPR)} = \dfrac{|FP|}{|FP| + |TN|}$

### Receiver Operating Characteristic (ROC)

An ROC curve plots TPR vs. FPR at different confidence thresholds.

### Area under the ROC curve (ROC AUC)

> From the ROC curve we calculate the ROC AUC metric by taking the integral using the trapezoidal rule formula.

Reviewer: "taking the integral" -> "estimating the integral"?

Reviewer: you aren't literally taking the integral in the equation, so you should delete that reference or change it to "estimating"

$$

> \sum_{i=1}^{|scores|} \frac{(TPR(scores_{i-1}) - TPR(scores_i)) \cdot (FPR(scores_{i-1}) - FPR(scores_i))}{2}

Reviewer: did you pull this formula from somewhere? it looks a little off to me, but I can't quite put my finger on why. I'd be interested to check my understanding by reading your source

Author: Its the trapezoidal rule formula modified for our use

Author: Ill work on it

$$

### References

> - [Classification: ROC Curve and AUC](https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc)

Reviewer: nit: can we delete this reference section since there's only one reference, and hyperlink to this page somewhere else in your above explaination?

Reviewer 2: i agree if there's just one link in reference section than the section is unnecessary

> ## Average Precision (AP) for Object Detection

Reviewer: remove "for Object Detection" since you're also talking about instance segmentation?

Reviewer: overall, this section was a bit hard to follow. could we add a signpost here at the top describing the main steps, then use the steps as the headers?

### Step 1: Use IOU to Match Predictions with Groundtruths
...

For object-detection and instance segmentation tasks, average-precision is calculated from the intersection-over-union (IoU) of geometric annotations.

### What is Intersection-over-Union (IoU)?

> IoU is the ratio of intersecting area over total area spanned by two geometries.

Reviewer: capitalize IOU everywhere for consistency

Author: Im fairly sure IoU is correct

Reviewer: see above

> IoU is the ratio of intersecting area over total area spanned by two geometries.

Reviewer: phrasing here reads a little awkwardly. what do you think about:

IOU is the ratio of the intersecting area of two geometries divided by the total area of those same geometries.

Author: IoU is the ratio of intersecting area over the joint area spanned by the two geometries.

$$Intersection \ over \ Union \ (IoU) = \dfrac{Area( prediction \cap groundtruth )}{Area( prediction \cup groundtruth )}$$

### Multiclass Precision and Recall

| Case | Description |
| :- | :- |

> | True Positive (TP) | Prediction meets IoU threshold requirements. |

Reviewer: "Prediction meets IoU threshold requirements." - can you clarify what this means?

| False Positive (FP) | Prediction fails IoU threshold requirements. |
| True Negative (TN) | Unused in multi-class evaluation.
| False Negative (FN) | No prediction exists. |

- $Precision = \dfrac{|TP|}{|TP| + |FP|} = \dfrac{\text{Number of True Predictions}}{|\text{Predictions}|}$

- $Recall = \dfrac{|TP|}{|TP| + |FN|} = \dfrac{\text{Number of True Predictions}}{|\text{Groundtruths}|}$

### Finding the best prediction for a groundtruth.

> To properly evaluate a detection we must first find the best matches of predictions to ground truths. We start by iterating over predictions by score from highest to lowest. For each prediction we assign the ground truth with the highest IoU value. Both the prediction and ground truth are now considered paired and removed from the pool of choices.

Reviewer: - comma after "detection"
- iterating over predictions by score from highest to lowest -> iterating over our predictions, starting with the highest scores first
- For each prediction we assign the ground truth with the highest IoU value. -> We pair each prediction with the ground truth that has the highest calculated IOU.

Note: For simplicity, the following algorithm assumes operation over the same datum and label.

$$
\begin{aligned}

> &\underline{\large\textbf{Algorithm 1} \hspace{0.5em} \text{Rank IOU's}} \\

Reviewer: I find this latex very confusing; I wonder if you could drastically shorten/simplify this section. some specific items I'm confused by:
- you offer more "Notes"s and summary details in the latex, rather than just including them in the preceeding text
- you're assigning variables rather than just using full names (like you did in the equations for previous sections)
- you're mixing python nomenclature (argsort) with (what I think is) pseudocode (e.g., foreach; for x do...); why not just make it as pythonic as possible without using variable names given we use python in our other examples?

\\
&8 \hspace{1.5em} I \gets \textbf{argsort}(scores) \\
&9 \hspace{1.5em} \textbf{foreach }i\text{ in }I\textbf{ do} \\
&10 \hspace{1em} | \quad p \gets predictions[i] \\
&11 \hspace{1em} | \quad \textbf{for }j=1\text{ to }k\textbf{ do} \\
&12 \hspace{1em} | \quad | \quad g \gets groundtruths[j] \\
&13 \hspace{1em} | \quad | \quad ious[j] \gets IoU(p, g) \\
&14 \hspace{1em} | \quad \textbf{end} \\
&15 \hspace{1em} | \quad J \gets \textbf{argsort}(ious) \\
&16 \hspace{1em} | \quad \textbf{foreach }j\text{ in }J\textbf{ do} \\
&17 \hspace{1em} | \quad | \quad \textbf{if }j\text{ not in }visited\_groundtruths \textbf{ then} \\
&18 \hspace{1em} | \quad | \quad | \quad visited.add(j) \\
&19 \hspace{1em} | \quad | \quad | \quad ranked\_ious.append(ious[j]) \\
&20 \hspace{1em} | \quad | \quad | \quad \textbf{break} \\
&21 \hspace{1em} | \quad | \quad \textbf{end} \\
&22 \hspace{1em} | \quad \textbf{end} \\
&23 \hspace{1em} \textbf{end} \\
\end{aligned}
$$

### Precision-Recall Curve

We can now compute the precision-recall curve using our previously ranked IoU's. We do this by iterating through the ranked IoU's and creating points cumulatively using recall and precision.

$$

> \begin{aligned}

Reviewer: same comments above apply here

&8 \hspace{1.5em} | \quad | \quad count\_tp \gets count\_tp + 1 \\
&9 \hspace{1.5em} | \quad \textbf{end} \\
&10 \hspace{1em} | \quad precision \gets count\_tp \mathrel{{/}} (i+1) \\
&11 \hspace{1em} | \quad recall \gets count\_tp \mathrel{{/}} n \\
&12 \hspace{1em} | \quad curve.append((recall, precision)) \\
&13 \hspace{1em} \textbf{end} \\
\end{aligned}
$$

### Calculating Average Precision

Average precision is defined as the integration of the precision-recall curve. However, due to the varying nature of datasets it has been shown that interpolating this curve with a fixed set of points helps to reduce inconsistencies between dataset splits. The defacto standard has been to use a 101-point interpolation of the precision-recall curve to compute this integral.

$$
AP = \frac{1}{101} \sum\limits_{r\in\{ 0, 0.01, \ldots , 1 \}}\rho_{interp}(r)
$$

$$
\rho_{interp} = \underset{\tilde{r}:\tilde{r} \ge r}{max \ \rho (\tilde{r})}
$$

### References
- [MS COCO Detection Evaluation](https://cocodataset.org/#detection-eval)
- [Mean Average Precision (mAP) Using the COCO Evaluator](https://pyimagesearch.com/2022/05/02/mean-average-precision-map-using-the-coco-evaluator/)

> ### Notes

Reviewer: can you move this up to below the original "Object Detection" section now that you've added more content between the two?

| False Negative (FN) | Prediction confidence score < threshold and is incorrect. |

- $\text{True Positive Rate (TPR)} = \dfrac{|TP|}{|TP| + |FN|}$

- $\text{False Positive Rate (FPR)} = \dfrac{|FP|}{|FP| + |TN|}$

These rates vary depending on a threshold so we can add a threshold parameter to be more explicit.

$$
TPR(threshold) = \dfrac{|TP(threshold)|}{|TP(threshold)| + |FN(threshold)|}
$$

$$
FPR(threshold) = \dfrac{|FP(threshold)|}{|FP(threshold)| + |TN(threshold)|}
$$

### Receiver Operating Characteristic (ROC)

An ROC curve plots TPR vs. FPR at different confidence thresholds.

In Valor, we use the confidence scores sorted in decreasing order as our thresholds.

$$
Point(score) = (FPR(score), \ TPR(score))
$$

### Area under the ROC curve (ROC AUC)

From the ROC curve we calculate the ROC AUC metric by approximating the integral using the trapezoidal rule formula.

$$
ROC AUC =  \sum_{i=1}^{|scores|} \frac{  \lVert Point(score_{i-1}) - Point(score_i) \rVert }{2}
$$

### References
- [Classification: ROC Curve and AUC](https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc)

## Average Precision (AP)

> For object-detection and instance segmentation tasks, average-precision is calculated from the intersection-over-union (IoU) of geometric annotations.

Reviewer 2: i prefer replacing "annotations" with "predictions and groundtruths". 

people usually don't think of an "annotation" as a prediction

### Multiclass Precision and Recall

> Geometric annotations use the ratio intersection-over-union (IoU) to calculate precision and recall. IoU is the ratio of intersecting area over the joint area spanned by the two geometries and is defined in the following equation.

Reviewer 2: ```suggestion
Tasks that predict geometries (such as object detection or instance segmentation) use the ratio intersection-over-union (IoU) to calculate precision and recall. IoU is the ratio of intersecting area over the joint area spanned by the two geometries and is defined in the following equation.
```
